xla/service/elemental_ir_emitter.cc

```diff
@@ -793,6 +793,123 @@ llvm::Value* EmitF8e4m3b11fnuzToF16(llvm::Value* f8_value,
   return f16_value;
 }
 
+absl::StatusOr<llvm::Value*> EmitF16ToF4e2m1fn(llvm::Value* f16_value,
+                                               llvm::IRBuilder<>* b) {
+  TF_ASSIGN_OR_RETURN(
+      llvm::Value * reduced_precision,
+      EmitReducePrecisionIR(
+          /*src_ty=*/F16, f16_value,
+          /*dest_exponent_bits=*/primitive_util::ExponentWidth(F4E2M1FN) + 1,
+          /*dest_mantissa_bits=*/primitive_util::SignificandWidth(F4E2M1FN) - 1,
+          /*quiet_nans=*/false, b));
+  llvm::Value* as_int16 = b->CreateBitCast(reduced_precision, b->getInt16Ty());
+  llvm::Value* as_int8 =
+      b->CreateTrunc(b->CreateLShr(as_int16, 9), b->getInt8Ty());
+
+  // Extract sign, exponent and mantissa from reduced precision value.
+  auto i8_const = [&](int val) {
+    return llvm::ConstantInt::get(b->getInt8Ty(), val);
+  };
+  llvm::Value* f4_sign = b->CreateLShr(as_int8, 6);
+  llvm::Value* f4_bits = b->CreateAnd(as_int8, i8_const(0x3F));
+  llvm::Value* f4_normal = b->CreateSub(f4_bits, i8_const(28));
+
+  // Special case for exponent overflow.
+  auto i16_const = [&](int val) {
+    return llvm::ConstantInt::get(b->getInt16Ty(), val);
+  };
+  llvm::Value* f16_bits = b->CreateAnd(
+      b->CreateBitCast(f16_value, b->getInt16Ty()), i16_const(0x7FFF));
+  llvm::Value* is_overflow =
+      b->CreateICmpUGE(f16_bits, i16_const(0x4700));                    // 7.0
+  llvm::Value* is_nan = b->CreateICmpUGT(f16_bits, i16_const(0x7C00));  // inf
+  llvm::Value* max_or_nan =
+      b->CreateSelect(is_nan, i8_const(0x8), i8_const(0x7));
```

Comment:
I find it slightly weird how a large F16 number gets turned into the max F4 value of 6, but an Inf F16 number gets turned to -0. I would expect a very big F16 and an Inf F16 get treated in the same way: either both get turned into 6 or both get turned into -0. But I'm fine with this current behavior if you think it's better than the alternative.

@mooskagh, let me know if you have an opinion here.

Reply:
Only NaN value gets turned to -0, the Inf value is converted to 6.0 (convert_test verifies that).